### ISSUE TITLE
Tooltip misbehaviour when using append-to and scrolling:

### ISSUE THREAD

**Issue by StefanCardnell** [wxsms/uiv#410]
**Describe the bug**
Placing a tooltip in an element other than `body` using the `append-to` property causes strange tooltip positioning while scrolling.

**To Reproduce**

Please see the minimal example:

https://codesandbox.io/s/uiv-bug-342-forked-4t8ic?fontsize=14&hidenavigation=1&theme=dark

Steps to reproduce the behaviour:
1. Add enough content in the document body so that it can be scrolled.
2. Create a button with a tooltip.
3. Alter the append-to of the tooltip to an element which is not `body`, in this case a div with an ID of `append_to_div`.
4. Hover over the button to see the tooltip appear above the button.
5. Scroll a little bit down, so that the button is still visible, and hover over the button again - the tooltip is now much higher than the button.

**Expected behavior**
The tooltip element should be appended to the `append_to_div` div element, not body, but should still be in the correct place regardless of scroll.

**Screenshots**

No scroll:

![image](https://user-images.githubusercontent.com/11324477/101636335-14abc580-3a23-11eb-9d00-5676fdead9f0.png)

With a little scroll:

![image](https://user-images.githubusercontent.com/11324477/101636364-1ffef100-3a23-11eb-8a08-65769c35fa50.png)

**Versions:**
 - Chrome version 87.
- UIV version 1.1.1

**Additional context**

I have examined the UIV code and can see why this happens. In the method `setTooltipPosition` the scrollTop of the `appendTo` container is used instead of the scrollTop of the `document`. As the container (the empty div) is empty, there is no scrolling and the subsequent calculation is off.

```javascript

export function setTooltipPosition (tooltip, trigger, placement, auto, appendTo, viewport) {
  if (!isElement(tooltip) || !isElement(trigger)) {
    return
  }
  const isPopover = tooltip && tooltip.className && tooltip.className.indexOf('popover') >= 0
  let containerScrollTop
  let containerScrollLeft
  if (!isExist(appendTo) || appendTo === 'body') {
    const doc = document.documentElement
    containerScrollLeft = (window.pageXOffset || doc.scrollLeft) - (doc.clientLeft || 0)
    containerScrollTop = (window.pageYOffset || doc.scrollTop) - (doc.clientTop || 0)
  } else {
    const container = getElementBySelectorOrRef(appendTo)
    containerScrollLeft = container.scrollLeft
    containerScrollTop = container.scrollTop
  }

```

I have also tried altering the `viewport` property to 1) the same append_to_div and 2) directly on body, neither worked.

**Motivation**

I am trying to append the tooltip to an element which is NOT `body` because the tooltip code uses `removeChild` when the tooltip hides. Unfortunately `removeChild` is quite slow when the parent element (in this case `body`) has many element, which my project does (it has thousands of \<td\> elements) - and it can make the page seem quite laggy when tooltips hide. Hence, I am trying to append the tooltip to a different element, but am facing this scrolling issue. 

If there is an alternate way to avoid the issue of slow `removeChild` on body, please let me know because then this scrolling issue won't matter too much for me.

My naive opinion would say that `setTooltipPosition` needs to consider the document scrolling even with a custom `appendTo`, but I don't know what other scenarios this would break. I appreciate this might be a difficult thing to alter.

**Comment by wxsms** [comment]
The behaviour was confirmed. However, when you use an `append-to` option, the scrolling position is decided by that element, it's by design. Consider tooltips inside scrolling modal.

I unserstand you case. maybe it's able to solve by splitting the `append-to` and `position-by` behaviour? like:

1. Add a new prop `position-by`.
2. When none of `append-to` and `position-by` is given, it will be `<body>` by default.
2. When only `append-to` is given, `position-by` is the same as `append-to`.
3. When `position-by` is given, tooltips will position by the given element.

This might be able to solve you problem.

But, are you sure that when you append the tooltip to an other element rather than body, your page becomes smooth? It sounds strange. It's your case base on IE? Can you provide a minimal reproduction for that case too? Thank you.

**Comment by StefanCardnell** [comment]
Hi @wxsms thanks for the detailed reply.

> But, are you sure that when you append the tooltip to an other element rather than body, your page becomes smooth? It sounds strange. It's your case base on IE? Can you provide a minimal reproduction for that case too? Thank you.

I've tried to make a minimal example for this situation here:

https://codesandbox.io/s/uiv-bug-342-forked-22h6o?file=/src/App.vue

This did not happen directly in codesandbox, I had to copy the code and run it directly in my browser (Chrome) for the issue to occur. If you open the model you might see that the opening stutters a bit. The stutter only occurs in Chrome and Edge, not Firefox, it could also be hardware dependent as well.

What happens is that:

1) The modal opens.
2) The tooltip on the button disappears as the modal opens.
3) As the tooltip disappears, it is removed by calling `removeChild` on the `body`.
4) `removeChild` on Chrome and Edge is quite slow, causing a recalculation of styles of all elements under the element for which `removeChild` was called. 
5) Subsequently you can see a slight "stutter" as the modal moves in to position as it opens.

It might be easier to open the Chrome dev tools performance profiler to analyse this. You'll see in the moment of the stutter, `removeFromDom` is being called. 

In my project I have thousands of \<td\>s and opening modals via buttons can seem stuttery for this reason. There are other things on the UI affected by a large `body` element, but modal opening was the most noticeable. My idea then was to append it to something else other than `body`.

> Add a new prop position-by.
> When none of append-to and position-by is given, it will be <body> by default.
> When only append-to is given, position-by is the same as append-to.
> When position-by is given, tooltips will position by the given element.
> This might be able to solve you problem.

I'd appreciate this solution, it would be ideal if `position-by` and `append-to` can both be used in a single directive, to avoid refactoring to use `tooltip` directly.

> Consider tooltips inside scrolling modal.

I tried to investigate this but I couldn't find a situation with scrolling modals and tooltips where the default `appendTo` setting didn't work perfectly already (which I guess makes sense, as modals are still within `body` and their position is still relative to `body`). I attach a minimal example with a scrolling modal, the tooltip seems to behave fine:

https://codesandbox.io/s/uiv-bug-342-forked-4zqhd?file=/src/App.vue


Either way, I appreciate the solution needs to be backwards compatible regardless, and changing the behaviour of the tooltip positioning with `append-to` might ruin a few people's day.

**Comment by wxsms** [comment]
> I'd appreciate this solution, it would be ideal if position-by and append-to can both be used in a single directive, to avoid refactoring to use tooltip directly.

Do you have any sugguestion to apply one more argument to the tooltip directive? As I see there can be only one argument, and is already taken by `append-to`.

**Comment by StefanCardnell** [comment]
@wxsms Apologies I just assumed it was possible. Having re-read the Vue documentation I can't see it either. It is fine as a separate attribute.

**Comment by wxsms** [comment]
Available in 1.2.0

**Comment by StefanCardnell** [comment]
Thanks @wxsms . I have tested and it works correctly. 

I've submitted a pull request so this can be specified in the directive as an object value. I previously didn't realise this functionality of using an object for the directive value was available before until I saw read the popover documentation.
